crates/sc-consensus-subspace-rpc/src/lib.rs

```diff
@@ -777,4 +759,62 @@ where
 
         Ok(None)
     }
+
+    async fn segment_headers(
+        &self,
+        segment_indexes: Vec<SegmentIndex>,
+    ) -> RpcResult<Vec<Option<SegmentHeader>>> {
+        if segment_indexes.len() > MAX_SEGMENT_HEADERS_PER_REQUEST {
+            error!(
+                "segment_indexes length exceed the limit: {} ",
+                segment_indexes.len()
+            );
+
+            return Err(JsonRpseeError::Custom(format!(
+                "segment_indexes length exceed the limit {MAX_SEGMENT_HEADERS_PER_REQUEST}"
+            )));
+        };
+
+        Ok(segment_indexes
+            .into_iter()
+            .map(|segment_index| self.segment_headers_store.get_segment_header(segment_index))
+            .collect())
+    }
+
+    async fn last_segment_headers(&self, limit: u64) -> RpcResult<Vec<Option<SegmentHeader>>> {
+        if limit as usize > MAX_SEGMENT_HEADERS_PER_REQUEST {
+            error!(
+                "Request limit ({}) exceed the server limit: {} ",
+                limit, MAX_SEGMENT_HEADERS_PER_REQUEST
+            );
+
+            return Err(JsonRpseeError::Custom(format!(
+                "Request limit ({}) exceed the server limit: {} ",
+                limit, MAX_SEGMENT_HEADERS_PER_REQUEST
+            )));
+        };
+
+        let runtime_api = self.client.runtime_api();
+        let best_hash = self.client.info().best_hash;
+
+        let last_segment_index = match runtime_api.history_size(best_hash) {
+            Ok(history_size) => history_size.segment_index().into(),
+            Err(error) => {
+                error!(?best_hash, "Failed to get history size: {}", error);
+
+                0u64 // starting segment index
+            }
+        };
+
+        let last_segment_headers = (0..=last_segment_index)
+            .rev()
+            .take(limit as usize)
+            .map(|segment_index| {
+                self.segment_headers_store
+                    .get_segment_header(segment_index.into())
+            })
+            .collect::<Vec<_>>();
```

Comment:
We should use `SegmentIndex` and not `u64`. Our APIs are designed to use it and it supports iteration as well, you can use `SegmentIndex::ZERO..=last_segment_index` and remove a bunch of `.into()`.